### PR TITLE
Generalize engine trait

### PR DESCRIPTION
**Pull request by rphmeier**
This introduces the parity-machine crate, which defines traits for blockchain headers and live blocks.

Consensus engines are intended to be generic over different kinds of state machines which fit the requirements they need. For example, ethash could potentially be run over any block chain type with a difficulty and balances (for block and uncle reward).

Verification and block population logic which was fragmented among engines has now been unified in two places: `ethcore/machine.rs` and `ethcore/verification/verification.rs`. This eliminates a common source of bugs stemming from protocol upgrades being implemented for some engines but not others.

`state`, `executive`, and `externalities` are now fully decoupled from consensus logic.

`ValidatorSet` based engines haven't been generalized yet, and currently are implemented only for the `EthereumMachine`. In a future PR, the `ValidatorSet` framework will be generalized to anything which supports a generalized notion of smart contracts and "signals" (implemented via ethabi and log events in Ethereum).

The snapshot module can is also a prime candidate for refactor. Currently there are two kinds of chunks: "block chunks" and "corroboration chunks". Corroboration chunks vary from engine to engine, and now we can make "block chunks" vary from machine to machine. This will give us secure warp sync with minimal effort for any new blockchains we create.

### CODE REVIEW

File: ethcore/res/ethereum/kovan.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
     "authorityRound": {
       "params": {
         "stepDuration": "4",
+   		"blockReward": "0x4563918244F40000",
```
Comment (arkpar) [inline ethcore/res/ethereum/kovan.json]: Indentation

File: machine/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+/// "seal" that indicates validity to a consensus engine.
+pub trait Header {
+	/// Cryptographic hash of the header, excluding the seal.
+    fn bare_hash(&self) -> H256;
+
+	/// Cryptographic hash of the header, including the seal.
+    fn hash(&self) -> H256;
+
+	/// Get a reference to the seal fields.
+    fn seal(&self) -> &[Vec<u8>];
```
Comment (arkpar) [inline machine/src/lib.rs]: Spaces are used instead of tabs here and above

File: ethcore/src/engines/null_engine.rs [hunk 3]
```diff
@@ -14,103 +14,92 @@
...
 
-	fn params(&self) -> &CommonParams {
-		&self.params
-	}
+	fn machine(&self) -> &M { &self.machine }
 
-	fn builtins(&self) -> &BTreeMap<Address, Builtin> {
-		&self.builtins
-	}
+	fn on_close_block(&self, block: &mut M::LiveBlock) -> Result<(), M::Error> {
```
Comment (arkpar) [inline ethcore/src/engines/null_engine.rs]: why not use `::engines::common::bestow_block_reward` here?

Comment (rphmeier) [reply ethcore/src/engines/null_engine.rs]: since it handles uncle reward as well and there is a test that depends on that behavior

**Comment by arkpar** [comment]
Looks good in general, mostly just code moved around.

**Comment by gavofyork** [comment]
@rphmeier ^^^

**Comment by 5chdn** [comment]
> Martin Holst Swende @holiman 13:55 (Gitter)
>
> It seems that #6591 blew the existing chain specifications out of the water; I'm getting quite a lot of errors when running parity on hive now. Is there some documentation about what's changed ?

I can help with docs if I get some pointers @rphmeier .

**Comment by bjornwgnr** [comment]
Are there any docs yet?

**Comment by 5chdn** [comment]
No but I can help drafting something if anyone asks.

**Comment by rphmeier** [comment]
Move `registrar`, `gasLimitBoundDivisor`, and `eip155Transition` from engine (Ethash/AuthorityRound/etc.) to general params section.

**Pull request merged**